### ISSUE TITLE
"user_id": user.name, AttributeError: 'NoneType' object has no attribute 'name'

### ISSUE THREAD

**Issue by mbeilin** [datopian/ckanext-authz-service#15]
Trying to download resource file in portal as anonymous (not authenticated in portal user) getting the following error:

```
2020-09-30 16:36:10,706 DEBUG [ckan.config.middleware.flask_app] Flask route match, endpoint: extstorage.download, args: {'filename': u'scmd_201901-1.csv', 'id': u'85d47f61-1032-4db9-b0f8-96a170afd269', 'resource_id': u'9914928f-99f9-4d4a-8b52-d612f6863dec'}, origin: extension
2020-09-30 16:36:10,791 DEBUG [ckan.config.middleware.pylons_app] Pylons route match: {'action': u'resource_download', 'controller': u'package', 'filename': u'scmd_201901-1.csv', 'id': u'85d47f61-1032-4db9-b0f8-96a170afd269', 'resource_id': u'9914928f-99f9-4d4a-8b52-d612f6863dec'} Origin: core
2020-09-30 16:36:10,792 DEBUG [ckan.config.middleware] Route support answers for GET /dataset/85d47f61-1032-4db9-b0f8-96a170afd269/resource/9914928f-99f9-4d4a-8b52-d612f6863dec/download/scmd_201901-1.csv: [(True, 'flask_app', 'extension'), (True, 'pylons_app', 'core')]
2020-09-30 16:36:10,792 DEBUG [ckan.config.middleware] Serving request via flask_app app
2020-09-30 16:36:10,862 DEBUG [ckan.logic] check access OK - resource_show user=
2020-09-30 16:36:10,862 DEBUG [ckan.logic] check access OK - package_show user=
2020-09-30 16:36:10,863 DEBUG [ckan.lib.search.query] Package query: {'q': u'name:"85d47f61-1032-4db9-b0f8-96a170afd269" OR id:"85d47f61-1032-4db9-b0f8-96a170afd269"', 'fq': 'site_id:"nhs-staging"', 'rows': 1, 'wt': 'json'}
2020-09-30 16:36:11,074 DEBUG [ckan.logic] check access OK - package_show user=
2020-09-30 16:36:11,075 DEBUG [ckan.lib.search.query] Package query: {'q': u'name:"85d47f61-1032-4db9-b0f8-96a170afd269" OR id:"85d47f61-1032-4db9-b0f8-96a170afd269"', 'fq': 'site_id:"nhs-staging"', 'rows': 1, 'wt': 'json'}
2020-09-30 16:36:11,319 DEBUG [ckan.logic] check access OK - package_show user=
2020-09-30 16:36:11,320 DEBUG [ckan.lib.search.query] Package query: {'q': 'name:"test-dataset-f11" OR id:"test-dataset-f11"', 'fq': 'site_id:"nhs-staging"', 'rows': 1, 'wt': 'json'}
2020-09-30 16:36:11,583 DEBUG [ckan.logic] check access OK - package_show user=
2020-09-30 16:36:11,595 DEBUG [ckan.logic] check access NotAuthorized - package_update user= "User  not authorized to edit package 85d47f61-1032-4db9-b0f8-96a170afd269"
2020-09-30 16:36:11,609 DEBUG [ckan.logic] check access NotAuthorized - package_update user= "User  not authorized to edit package 85d47f61-1032-4db9-b0f8-96a170afd269"
2020-09-30 16:36:11,609 DEBUG [ckan.logic] check access NotAuthorized - dataset_purge user= "<function dataset_purge at 0x7f33a3d625f0> requires an authenticated user"
2020-09-30 16:36:11,609 DEBUG [ckan.logic] check access NotAuthorized - package_delete user= "<function package_delete at 0x7f33a3d62578> requires an authenticated user"
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/paste/httpserver.py", line 1068, in process_request_in_thread
    self.finish_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 331, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python2.7/SocketServer.py", line 652, in __init__
    self.handle()
  File "/usr/lib/python2.7/site-packages/paste/httpserver.py", line 442, in handle
    BaseHTTPRequestHandler.handle(self)
  File "/usr/lib/python2.7/BaseHTTPServer.py", line 340, in handle
    self.handle_one_request()
  File "/usr/lib/python2.7/site-packages/paste/httpserver.py", line 437, in handle_one_request
    self.wsgi_execute()
  File "/usr/lib/python2.7/site-packages/paste/httpserver.py", line 287, in wsgi_execute
    self.wsgi_start_response)
  File "/srv/app/src/ckan/ckan/config/middleware/__init__.py", line 202, in __call__
    return self.apps[app_name](environ, start_response)
  File "/usr/lib/python2.7/site-packages/repoze/who/middleware.py", line 86, in __call__
    app_iter = app(environ, wrapper.wrap_start_response)
  File "/usr/lib/python2.7/site-packages/webob/dec.py", line 147, in __call__
    resp = self.call_func(req, *args, **self.kwargs)
  File "/usr/lib/python2.7/site-packages/webob/dec.py", line 208, in call_func
    return self.func(req, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/fanstatic/publisher.py", line 234, in __call__
    return request.get_response(self.app)
  File "/usr/lib/python2.7/site-packages/webob/request.py", line 1053, in get_response
    application, catch_exc_info=False)
  File "/usr/lib/python2.7/site-packages/webob/request.py", line 1022, in call_application
    app_iter = application(self.environ, start_response)
  File "/usr/lib/python2.7/site-packages/webob/dec.py", line 147, in __call__
    resp = self.call_func(req, *args, **self.kwargs)
  File "/usr/lib/python2.7/site-packages/webob/dec.py", line 208, in call_func
    return self.func(req, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/fanstatic/injector.py", line 54, in __call__
    response = request.get_response(self.app)
  File "/usr/lib/python2.7/site-packages/webob/request.py", line 1053, in get_response
    application, catch_exc_info=False)
  File "/usr/lib/python2.7/site-packages/webob/request.py", line 1022, in call_application
    app_iter = application(self.environ, start_response)
  File "/usr/lib/python2.7/site-packages/flask/app.py", line 1997, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/lib/python2.7/site-packages/beaker/middleware.py", line 156, in __call__
    return self.wrap_app(environ, session_start_response)
  File "/usr/lib/python2.7/site-packages/flask/app.py", line 1985, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/lib/python2.7/site-packages/flask/app.py", line 1540, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/lib/python2.7/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/lib/python2.7/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/lib/python2.7/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/lib/python2.7/site-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/lib/python2.7/site-packages/flask_debugtoolbar/__init__.py", line 125, in dispatch_request
    return view_func(**req.view_args)
  File "/srv/app/ckanext-external-storage/ckanext/external_storage/blueprints.py", line 30, in download
    return call_download_handlers(resource, package, filename)
  File "/srv/app/ckanext-external-storage/ckanext/external_storage/download_handler.py", line 38, in call_download_handlers
    response = plugin.resource_download(resource, package, filename)
  File "/srv/app/ckanext-external-storage/ckanext/external_storage/plugin.py", line 60, in resource_download
    return download_handler(resource, package, filename)
  File "/srv/app/ckanext-external-storage/ckanext/external_storage/download_handler.py", line 54, in download_handler
    resource_download_spec = tk.get_action('get_resource_download_spec')(context, data_dict)
  File "/srv/app/src/ckan/ckan/logic/__init__.py", line 464, in wrapped
    result = _action(context, data_dict, **kw)
  File "/srv/app/ckanext-external-storage/ckanext/external_storage/actions.py", line 22, in get_resource_download_spec
    client = get_lfs_client(context, resource)
  File "/srv/app/ckanext-external-storage/ckanext/external_storage/actions.py", line 38, in get_lfs_client
    return context.get('lfs_client', LfsClient(helpers.server_url(), _get_authz_token(context, resource)))
  File "/srv/app/ckanext-external-storage/ckanext/external_storage/actions.py", line 72, in _get_authz_token
    authz_result = authorize(context, {"scopes": [scope]})
  File "/srv/app/src/ckan/ckan/logic/__init__.py", line 464, in wrapped
    result = _action(context, data_dict, **kw)
  File "/srv/app/src/ckanext-authz-service/ckanext/authz_service/actions.py", line 38, in authorize
    return {"user_id": user.name,
AttributeError: 'NoneType' object has no attribute 'name'
```

**Comment by shevron** [comment]
I've been out of the code for a while now, but IIRC there is no handling for anonymous access in ckanext-authz-service at all right now (this was never implemented because up until now no users which have used this extension required it). It *does* make sense to add it. 

The logic for supporting anonymous access *should be* quite simple: once a "user" has been identified as anonymous, it should be up to the CKAN configuration and built-in authorization functions to decide whether they get access or not. 

Basically, anonymous access support is missing and needs to be implemented.
